util/bloom/xor.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package bloom
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+
+	"github.com/algorand/xorfilter"
+)
+
+// XorBuilder is a local alias for xorfilter.Builder
+type XorBuilder = xorfilter.Builder
+
+// XorFilter is a faster more efficient alternative to a Bloom filter
+// An XorFilter object can be used as is or with optional adittional setup.
+type XorFilter struct {
+	//xor     *xorfilter.Xor8
+	xor     *xorfilter.Xor32
+	holding map[uint64]bool
+
+	b *XorBuilder
+}
+
+// NewXor returns an XorFilter with an internal map created with a size hint and an optional *XorBuilder (may be nil)
+// The Builder is not thread safe and should only be used by one thread at a time.
+func NewXor(hint int, builder *XorBuilder) *XorFilter {
+	return &XorFilter{
+		holding: make(map[uint64]bool, hint),
+		b:       builder,
+	}
+}
+
+// Set adds the value to the filter.
+func (xf *XorFilter) Set(x []byte) {
+	if xf.holding == nil {
+		xf.holding = make(map[uint64]bool)
+	}
+	k := binary.BigEndian.Uint64(x)
+	xf.holding[k] = true
+}
+
+// Test checks whether x is present in the filter.
+// May return (rare) erroneous true values, but false is precise.
+func (xf *XorFilter) Test(x []byte) bool {
+	k := binary.BigEndian.Uint64(x)
+	if xf.holding != nil {
+		return xf.holding[k]
+	}
+	if xf.xor != nil {
+		return xf.xor.Contains(k)
+	}
+	return false
+}
+
+// MarshalBinary implements encoding.BinaryMarshaller interface
+func (xf *XorFilter) MarshalBinary() ([]byte, error) {
+	if len(xf.holding) != 0 {
+		keys := make([]uint64, len(xf.holding))
+		pos := 0
+		for k := range xf.holding {
+			keys[pos] = k
+			pos++
+		}
+		var err error
```

Comment:
I think that if you're already doing a dedup, you'd want to `keys = keys[:pos]` so that you won't be passing empty duplicated entries to Populate, right ?

Reply:
pretty sure keys was allocated to the precise length, no extra positions.